### PR TITLE
updated the geth-with-protocol docker to use the confluence branch an…

### PR DESCRIPTION
**Pull request by RiccardoBiosas**
**What does this pull request do?**
Updates the dockerfile and the build script of the protocol-with-geth docker in order to run the confluence version of the livepeer protocol with hardhat rather than truffle

**Does this pull request close any open issues?**
https://github.com/livepeer/go-livepeer/issues/2315

**TODO**
This line needs to be updated to point to the correct branch once the the related [PR](https://github.com/livepeer/protocol/pull/557) on the livepeer-protocol repo will be merged
https://github.com/livepeer/docker-livepeer/blob/ee511758cf552c6022cc61a875a8fb6de81e0f9b/geth-with-protocol/build-protocol.sh#L3-L4

### CODE REVIEW

File: geth-with-protocol/Dockerfile [hunk 1]
```diff
@@ -9,7 +9,7 @@ RUN echo "" > /geth/password.txt
 RUN geth --datadir $gethRoot init /geth/genesis.json
 RUN cp /geth/keys/* $gethRoot/keystore/
 
-FROM node:10-alpine as builder-node
+FROM node:14.4.0-alpine as builder-node
```
Comment (0xcadams) [inline geth-with-protocol/Dockerfile:12]: Was this `node` version chosen for compatibility? It looks like the latest v14 release is:

https://github.com/nodejs/node/releases/tag/v14.19.1

**Review by leszko — CHANGES_REQUESTED** [review]
@RiccardoBiosas Thanks for the PR, I really like that you changed Truffle to Hardhat!

A few comments:
1. I've build the image and trying to use it with the [Livepeer Devtool](https://github.com/livepeer/go-livepeer/tree/master/cmd/devtool), but when running the following command, it hangs.

```
$ go run cmd/devtool/devtool.go setup transcoder
...
I0509 18:32:22.163257   45207 devtool.go:263] Waiting will first round ended.
I0509 18:32:26.175735   45207 devtool.go:263] Waiting will first round ended.
I0509 18:32:30.178201   45207 devtool.go:263] Waiting will first round ended.
I0509 18:32:34.186991   45207 devtool.go:263] Waiting will first round ended.
```

I remember that in the old geth image, the round was way shorter than in the real chain. Maybe that's a clue. Could you try running devtool commands and check what's the issue?

2. After fixed, could you build the image and push to `livepeer/livepeer:confluence`?
3. Could you update the [related doc in docker-livepeer](https://github.com/livepeer/docker-livepeer/blob/master/geth-with-protocol/README.md)?
4. Could you update the [related doc in go-livepeer](https://github.com/livepeer/go-livepeer/blob/master/cmd/devtool/README.md)?

**Comment by RiccardoBiosas** [comment]
> @RiccardoBiosas Thanks for the PR, I really like that you changed Truffle to Hardhat!
> 
> A few comments:
> 
> 1. I've build the image and trying to use it with the [Livepeer Devtool](https://github.com/livepeer/go-livepeer/tree/master/cmd/devtool), but when running the following command, it hangs.
> 
> ```
> $ go run cmd/devtool/devtool.go setup transcoder
> ...
> I0509 18:32:22.163257   45207 devtool.go:263] Waiting will first round ended.
> I0509 18:32:26.175735   45207 devtool.go:263] Waiting will first round ended.
> I0509 18:32:30.178201   45207 devtool.go:263] Waiting will first round ended.
> I0509 18:32:34.186991   45207 devtool.go:263] Waiting will first round ended.
> ```
> 
> I remember that in the old geth image, the round was way shorter than in the real chain. Maybe that's a clue. Could you try running devtool commands and check what's the issue?
> 
> 2. After fixed, could you build the image and push to `livepeer/livepeer:confluence`?
> 3. Could you update the [related doc in docker-livepeer](https://github.com/livepeer/docker-livepeer/blob/master/geth-with-protocol/README.md)?
> 4. Could you update the [related doc in go-livepeer](https://github.com/livepeer/go-livepeer/blob/master/cmd/devtool/README.md)?

I think I faced the same issue and it should be because the confluence version of the livepeer protocol deploys the test contract AdjustableRoundsManager - rather than the actual RoundsManager - when the network is not one of [the live networks](https://github.com/livepeer/protocol/blob/confluence/deploy/deploy_contracts.ts#L26-L33).
AdjustableRoundsManager overrides the [`blockNum()` function of the RoundsManager contract](https://github.com/livepeer/protocol/blob/confluence/contracts/rounds/RoundsManager.sol#L122-L124), which forces you to [manually set the block number](https://github.com/livepeer/protocol/blob/confluence/contracts/rounds/AdjustableRoundsManager.sol#L12-L14) to initialize a new round.
The way I got around it was to open a geth console, initialize the AdjustableRoundsManager and call `blockNum`, i.e.:
```
const adjustableRoundsManager = await ethers.getContractAt("AdjustableRoundsManager", 'addressOfAdjustableRoundsManager')
await adjustableRoundsManager.setBlockNum(roundsLengthN)
```
I guess the rationale was to avoid waiting N rounds in a test environment but personally I find it a bit cumbersome, perhaps we could use the actual RoundsManager to test go-livepeer (assuming that's the cause of your issue).

**Comment by leszko** [comment]
> I guess the rationale was to avoid waiting N rounds in a test environment but personally I find it a bit cumbersome, perhaps we could use the actual RoundsManager to test go-livepeer (assuming that's the cause of your issue).

Yes, I agree with you. Maybe we don't need `AdjustableRoundsManager` at all. @yondonfu what do you think? Can we just use the "real" `RoundsManager` in the tests?

**Comment by yondonfu** [comment]
> Can we just use the "real" RoundsManager in the tests?

Yes I think that's fine and actually the way the current geth-with-protocol Docker image was setup since it was built the protocol repo from an old commit hash where the "lpTestNet" network is [considered a "live network"](https://github.com/livepeer/protocol/blob/d8e3523e3779a7f339b94159ad5d36ce428799cc/utils/contractDeployer.js#L66) and the build script [used the lpTestNet network when deploying](https://github.com/livepeer/docker-livepeer/blob/e9dc8ad6949f0e4301e859718d505ca37fc94b5a/geth-with-protocol/build-protocol.sh#L27). So, making sure to use the "real" RoundsManager in this update would align with the previous behavior (it just happens to be the case that the lpTestNet network was removed at some point so the current tip of the protocol repo Confluence branch will use the AdjustableRoundsManager when building the protocol repo for this image).

File: geth-with-protocol/Dockerfile (hunk 1 above)
Comment (0xcadams) [reply geth-with-protocol/Dockerfile:12]: Sounds good, we may want to change this to `14-alpine` to reduce maintenance

**Comment by leszko** [comment]
> > Can we just use the "real" RoundsManager in the tests?
> 
> Yes I think that's fine and actually the way the current geth-with-protocol Docker image was setup since it built the protocol repo from an old commit hash where the "lpTestNet" network was [considered a "live network"](https://github.com/livepeer/protocol/blob/d8e3523e3779a7f339b94159ad5d36ce428799cc/utils/contractDeployer.js#L66) and the build script [used the lpTestNet network when deploying](https://github.com/livepeer/docker-livepeer/blob/e9dc8ad6949f0e4301e859718d505ca37fc94b5a/geth-with-protocol/build-protocol.sh#L27). So, making sure to use the "real" RoundsManager in this update would align with the previous behavior (it just happens to be the case that the lpTestNet network was removed at some point so the current tip of the protocol repo Confluence branch will use the AdjustableRoundsManager when building the protocol repo for this image).

Ok, good, then I suggest we just use `RoundsManager` and just remove `AdjustableRoundsManager` if it's not used anywhere.

**Comment by RiccardoBiosas** [comment]
> > > Can we just use the "real" RoundsManager in the tests?
> > 
> > 
> > Yes I think that's fine and actually the way the current geth-with-protocol Docker image was setup since it built the protocol repo from an old commit hash where the "lpTestNet" network was [considered a "live network"](https://github.com/livepeer/protocol/blob/d8e3523e3779a7f339b94159ad5d36ce428799cc/utils/contractDeployer.js#L66) and the build script [used the lpTestNet network when deploying](https://github.com/livepeer/docker-livepeer/blob/e9dc8ad6949f0e4301e859718d505ca37fc94b5a/geth-with-protocol/build-protocol.sh#L27). So, making sure to use the "real" RoundsManager in this update would align with the previous behavior (it just happens to be the case that the lpTestNet network was removed at some point so the current tip of the protocol repo Confluence branch will use the AdjustableRoundsManager when building the protocol repo for this image).
> 
> Ok, good, then I suggest we just use `RoundsManager` and just remove `AdjustableRoundsManager` if it's not used anywhere.

It's being used in the protocol unit/integration tests (see [here](https://github.com/livepeer/protocol/blob/confluence/test/integration/TicketFrontRun.js#L84-L87)). However I can make sure that for the purpose of the e2e testing on the geth environment we use  the RoundsManager

**Review by leszko — APPROVED** [review]
Looks great 👍 Good work @RiccardoBiosas 🔥 

Remember to:
1. First merge the related protocol PR and update `TODO` here to use the `confluece` branch instead of your branch
2. Squash commits (or squash and merge) to merge this PR
3. After merge, check if the image `livepeer/geth-with-livepeer-protocol:confluence` is pushed into Docker Hub

**Pull request merged**